gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencyMerger.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.dist;
+
+import static com.palantir.logsafe.Preconditions.checkArgument;
+
+import com.palantir.logsafe.SafeArg;
+import com.palantir.sls.versions.OrderableSlsVersion;
+import com.palantir.sls.versions.VersionComparator;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+public final class RecommendedProductDependencyMerger {
+    private RecommendedProductDependencyMerger() { }
+
+    public static RecommendedProductDependency mergeRecommendedProductDependencies(
+            RecommendedProductDependency dep1, RecommendedProductDependency dep2) {
+        // Ensure they are valid
+        Arrays.asList(dep1, dep2).forEach(RecommendedProductDependency::isValid);
+        checkArgument(
+                dep1.getProductGroup().equals(dep2.getProductGroup()),
+                "Product groups differ",
+                SafeArg.of("dep1ProductGroup", dep1.getProductGroup()),
+                SafeArg.of("dep2ProductGroup", dep2.getProductGroup()));
+        checkArgument(
+                dep1.getProductName().equals(dep2.getProductName()),
+                "Product names differ",
+                SafeArg.of("dep1ProductName", dep1.getProductName()),
+                SafeArg.of("dep2ProductName", dep2.getProductName()));
+
+        VersionComparator versionComparator = VersionComparator.INSTANCE;
+        OrderableSlsVersion minimumVersion = Collections.max(
+                Arrays.asList(
+                        OrderableSlsVersion.valueOf(dep1.getMinimumVersion()),
+                        OrderableSlsVersion.valueOf(dep2.getMinimumVersion())),
+                versionComparator);
+
+        MaximumVersionComparator maximumVersionComparator = MaximumVersionComparator.INSTANCE;
+        Optional<MaximumVersion> maximumVersion = Stream
+                .of(dep1.getMaximumVersion(), dep2.getMaximumVersion())
```

Comment:
kinda funky to use Stream.of here but Collections.max(Arrays.asList()) above... pick one and run with it?